### ISSUE TITLE
Improve unification for injective symbols

### ISSUE THREAD

**Issue by 01mf02** [Deducteam/lambdapi#245]
When using implicit arguments, it currently seems that one sometimes has to give arguments explicitly although they should be able to be inferred automatically. Consider the following example:

~~~
symbol const all : ∀ {A : Type}, eta ((A > o) > o)
rule eps (@all &a &b) → ∀ (x : eta &a), eps (&b x)

// Leibniz equality
definition eq :
  ∀ {A : Type}, eta (A > A > o) ≔
  λ A x y, @all (A > o) (λ p, p x --> p y)
~~~

In the definition of `eq`, one can infer that `x` and `y` are of type `A` (because of the type annotation `(A > A > o)`, and therefore, `p` has to have type `(A > o)`, making the explicit type annotation `(A > o)` superfluous. Alas, when we omit it:

~~~
// Leibniz equality, second try
definition eq_implicit :
  ∀ {A : Type}, eta (A > A > o) ≔
  λ A x y, all (λ p, p x --> p y)
~~~

We obtain an error:

    [implicit-ex.lp, 27:0-29:33] Term [λA x y, all ?26[A,x,y] (λp, p x --> p y)] does not have type ∀(A:Type), eta (A > (A > o))].

**Comment by 01mf02** [comment]
The full example:

~~~
symbol const Type : TYPE
symbol injective eta : Type ⇒ TYPE

// function type
symbol const Ar : Type ⇒ Type ⇒ Type
set infix right 6 ">" ≔ Ar
rule eta (&a > &b) → eta &a ⇒ eta &b

symbol const i : Type
symbol const o : Type

symbol eps : eta o ⇒ TYPE

symbol const imp : eta (o > o > o)
rule eps (imp &a &b) → eps &a ⇒ eps &b
set infix right 6 "-->" ≔ imp

symbol const all : ∀ {A : Type}, eta ((A > o) > o)
rule eps (@all &a &b) → ∀ (x : eta &a), eps (&b x)

// Leibniz equality
definition eq :
  ∀ {A : Type}, eta (A > A > o) ≔
  λ A x y, @all (A > o) (λ p, p x --> p y)

// Leibniz equality, second try (currently does not work)
definition eq_implicit :
  ∀ {A : Type}, eta (A > A > o) ≔
  λ A x y, all (λ p, p x --> p y)
~~~

**Comment by fblanqui** [comment]
It's possible to check this file by adding the following declarations after the definition of all:
```
set builtin "T" ≔ eta
set builtin "P" ≔ eta
set builtin "all" ≔ all
```

**Comment by fblanqui** [comment]
Note that this inference is not trivial. It boils down to finding a term X such that eta X = eta A ⇒ eta o modulo beta-reduction and rewriting. There is a unique solution in this case because eta is injective. 

For the moment, the unification algorithm does not handle arbitrary injective symbols but only the symbol mapped to "P" in a (not documented) set builtin declaration  (*) where, normally: 

U : TYPE // is the LP type of L-types where L is the logical system you encode in LP
T : U ⇒ TYPE // is the LP function interpreting L-types as LP types
Prop : TYPE // is the LP type of L-propositions
P : Prop ⇒ TYPE // is the LP function interpreting L-propositions as LP types
all: ∀ a : U, (T a ⇒ Prop) ⇒ Prop // represents typed universal quantification in L

In your example, we should normally take:
U = Type
T = eta
Prop = eta o
P = eps
all = all

But this is not sufficient to solve your problem because what you have to inverse is not P but T. Hence, the trick mentioned in the previous comment:
T = eta
P = eta
all = Ar
which is incorrect wrt typing but does the job, more or less by chance. This should be fixed somehow.

(*) The builtins "T" and "P" are also used in equality-related tactics (see https://github.com/Deducteam/lambdapi/blob/master/doc/sections/commands.md). "P" is also used in Yacine's why3 branch.

**Comment by fblanqui** [comment]
Problem fixed in 5f9b90411374b98c28accbd2eb6d77ae40eee639. There is no need for builtins anymore.

**Comment by fblanqui** [comment]
Note that `type λ A x y, all (λ p, p x --> p y)` fails because, in this case, one has to find X such that eta X = Z ⇒ eta o, where Z is the type of x, which is unknown also, but this is not possible.

**Comment by 01mf02** [comment]
Thank you, Frédéric.
Your new solution looks much better.
I still have found a new case which Dedukti refuses to typecheck.
The example below also contains some other tests which previously used not to work, but which work now. I include them for completeness (and potentially for the test suite).

~~~
symbol const Type : TYPE
symbol injective eta : Type ⇒ TYPE

// function type
symbol const Ar : Type ⇒ Type ⇒ Type
set infix right 6 ">" ≔ Ar
rule eta (&a > &b) → eta &a ⇒ eta &b

symbol const i : Type
symbol const o : Type

symbol injective eps : eta o ⇒ TYPE

symbol const imp : eta (o > o > o)
rule eps (imp &a &b) → eps &a ⇒ eps &b
set infix right 6 "-->" ≔ imp

symbol const all : ∀ {A : Type}, eta ((A > o) > o)
rule eps (@all &a &b) → ∀ (x : eta &a), eps (&b x)

definition id : ∀ {A : Type}, eta (A > A) ≔ λ A x, x

symbol const zero : eta i
symbol const succ : eta (i > i)

// works
type id zero
type @id (i > i) succ
type id succ

definition true : eta o ≔ all (λ p, p --> p)
definition trueI : eps true ≔ λ P p, p
definition imp_refl : ∀ {P : eta o}, eps (P --> P) ≔ λ P prf, prf
definition true_imp_true : eps (true --> true) ≔ @imp_refl true

// works
type @imp_refl true trueI
type @imp_refl (true --> true) true_imp_true

// does not work
//type imp_refl trueI
//type imp_refl true_imp_true
~~~

To compare, a similar Coq version:

~~~ coq
Axiom imp_refl : forall {A : Prop}, A -> A.
Definition true_imp_true : True -> True := @imp_refl True.

Type @imp_refl (True -> True) true_imp_true.
Type imp_refl true_imp_true.


Definition id : forall {A : Type}, A -> A := fun A x => x.

Axiom A : Type.
Axiom Zero : A.
Axiom Succ : A -> A.

Type id Zero.
Type id Succ.
~~~

**Comment by 01mf02** [comment]
> Note that `type λ A x y, all (λ p, p x --> p y)` fails because, in this case, one has to find X such that eta X = Z ⇒ eta o, where Z is the type of x, which is unknown also, but this is not possible.

Yes, is clear.

**Comment by 01mf02** [comment]
I have found yet another problem: When I modify my original example as follows:

~~~
[...]

// Leibniz equality
definition eq :
  ∀ {A : Type}, eta (A > A > o) ≔
  λ A x y, all (λ p, p x --> p y)
set infix right 7 "≡" ≔ eq

// reflexivity of equality
definition eq_refl :
  ∀ {A : Type}, eps (@all A (λ x, x ≡ x)) ≔
  λ A x q H, H

// symmetry of equality
definition eq_sym :
  ∀ {A : Type} (x y : eta A),
  eps (x ≡ y --> y ≡ x) ≔
  λ A x y H, H (λ y, y ≡ x) (eq_refl x)
~~~

Then the proof of `eq_sym` no longer works.
However, if I use the original version of `eq`, then Dedukti typechecks the example.

**Comment by 01mf02** [comment]
(I also added `injective` to the declaration of `eps` in my original example.)

**Comment by fblanqui** [comment]
Fixed in b8c056b3fca71a8f314d39a11a8a2ddc84c663e5.

**Comment by 01mf02** [comment]
Even with the fix in b8c056b, there remains the issue with `eq_sym` I already mentioned. The error message is:

    [λx0, o] and [o] are not convertible.

For reference, here is the full example:

~~~
symbol const Type : TYPE
symbol injective eta : Type ⇒ TYPE

// function type
symbol const Ar : Type ⇒ Type ⇒ Type
set infix right 6 ">" ≔ Ar
rule eta (&a > &b) → eta &a ⇒ eta &b

symbol const i : Type
symbol const o : Type

symbol eps : eta o ⇒ TYPE

symbol const imp : eta (o > o > o)
rule eps (imp &a &b) → eps &a ⇒ eps &b
set infix right 6 "-->" ≔ imp

symbol const all : ∀ {A : Type}, eta ((A > o) > o)
rule eps (@all &a &b) → ∀ (x : eta &a), eps (&b x)


// Leibniz equality
definition eq :
  ∀ {A : Type}, eta (A > A > o) ≔
// works
  λ A x y, @all (A > o) (λ p, p x --> p y)
// fails when typechecking `eq_sym`
//  λ A x y, all (λ p, p x --> p y)
set infix right 7 "≡" ≔ eq

// reflexivity of equality
definition eq_refl :
  ∀ {A : Type}, eps (@all A (λ x, x ≡ x)) ≔
  λ A x q H, H

// symmetry of equality
definition eq_sym :
  ∀ {A : Type} (x y : eta A),
  eps (x ≡ y --> y ≡ x) ≔
  λ A x y H, H (λ y, y ≡ x) (eq_refl x)


// functional extensionality
symbol const func_ext :
  ∀ {A B : Type} (f g : eta (A > B)),
// works
  eps ((all (λ x, f x ≡ g x)) --> (@eq (A > B) f g))
// fails
//  eps ((all (λ x, f x ≡ g x)) --> (eq f g))
~~~

To the example, I added yet another case, `func_ext`, where omitting an implicit argument fails with the error message:

    [B] and [λ_, B] are not convertible.

Other theorems where I tried to eliminate the implicit arguments expose similar error messages.

**Comment by fblanqui** [comment]
This is a bug in my new code. I'm going to try to fix it ASAP.

**Comment by fblanqui** [comment]
Fixed in 74ae2e2125f7565dd4dc1024f99bc92854bac5ca.
